contracts/SolutionSubmitter.sol

```diff
@@ -0,0 +1,73 @@
+pragma solidity ^0.5.0;
+
+import "./BatchExchange.sol";
+import "@openzeppelin/contracts/ownership/Ownable.sol";
+
+interface GasToken {
+    function freeUpTo(uint256 value) external returns (uint256 freed);
+}
+
+/** @title SolutionSubmitter - A simple wrapper contract allowing to submit settlements
+ * to the BatchExchange contract while saving gas via the GST2 or CHI GasToken
+ */
+contract SolutionSubmitter is Ownable {
+    BatchExchange public exchange;
+    GasToken public gasToken;
+    uint256 public gasPriceThreshold;
+
+    constructor(
+        BatchExchange _exchange,
+        GasToken _gasToken,
+        uint256 _gasPriceThreshold
+    ) public {
+        exchange = _exchange;
+        setGasToken(_gasToken);
+        setGasPriceThreshold(_gasPriceThreshold);
+    }
+
+    function setGasToken(GasToken _gasToken) public {
+        gasToken = _gasToken;
+    }
+
+    function setGasPriceThreshold(uint256 _gasPriceThreshold) public {
+        gasPriceThreshold = _gasPriceThreshold;
+    }
+
+    /**
+     * Allow the owner to execute arbitrary code (e.g. transfer out solution fee rewards from BatchExchange)
+     */
+    function execute(address target, bytes calldata data) external onlyOwner returns (bool, bytes memory) {
+        // solium-disable-next-line security/no-low-level-calls
+        return target.call(data);
+    }
+
+    /**
+     * Wrapper around actual solution submission that uses gas tokens for discounts
+     */
+    function submitSolution(
+        uint32 batchId,
+        uint256 claimedObjectiveValue,
+        address[] memory owners,
+        uint16[] memory orderIds,
+        uint128[] memory buyVolumes,
+        uint128[] memory prices,
+        uint16[] memory tokenIdsForPrice
+    ) public gasDiscounted returns (uint256) {
+        return exchange.submitSolution(batchId, claimedObjectiveValue, owners, orderIds, buyVolumes, prices, tokenIdsForPrice);
+    }
+
+    /**
+     * Modifier to invoke original method and freeing up to half the maximum gas refund in gas tokens.
+     * Logic adjusted from https://1inch-exchange.medium.com/1inch-introduces-chi-gastoken-d0bd5bb0f92b
+     */
+    modifier gasDiscounted {
```

Comment:
Interesting that you are working with a set gasPriceThreshold.

Wouldn't it be better to work with a modifier `gasDiscounted(bool activated)`, where the activated parameter is part of the function signature. Then we can easily activate or deactivate the gas discounting in each call without setting this gasPriceThreshold. 
Also, it should be cheaper gas wise...

Reply:
You are right that the proposed solution requires an SLOAD on every solution submission. The reason I didn't go with an `activated` argument in the function signature was to leave the function signature exactly as the one on the original contract (thus the driver code doesn't even need to know it's using a gas token wrapper).

Otherwise, the driver code could directly have the logic of whether to hit the wrapper contract or BatchExchange directly (no need to defer that to solidity). Small disadvantage there could is that then the same solver (e.g. prod standard staging) could submit solutions from two different `msg.sender` addresses.

Reply:
One solution would be to use `immutable` (so it doesn't incur an `SLOAD` but instead works like a `constant`). The downside being you would need to redeploy a new contract to update the gas token and threshold. If this is not expected to be done often, then it might be worth it?

Reply:
I think in the experimentation phase, I'd prefer to change the values without redeploying the contract (which would also mean we need to change the services again to use a different wrapper).